### PR TITLE
Feature/piv 31/issue jwt

### PR DESCRIPTION
**Pull request by boyan-d**

### CODE REVIEW

File: database/users/db.go [hunk 1]
```diff
@@ -25,12 +26,12 @@ const (
 	updateUserProfileQuery    = `UPDATE "user" SET profile = $2 WHERE user_id = $1;`
 
 	upsertUserQuery = `INSERT INTO "user"
```
Comment (cnaize) [inline database/users/db.go:28]: where are you using new field? I don't see any places where this additional parameter used

File: universe/node/api_auth.go [hunk 2]
```diff
@@ -164,3 +174,95 @@ func (n *Node) apiGenToken(c *gin.Context) {
 
 	c.JSON(http.StatusOK, out)
 }
+
+func (n *Node) apiGenerateGuestToken(c *gin.Context) {
```
Comment (cnaize) [inline universe/node/api_auth.go:178]: no swagger

File: universe/node/api_auth.go [hunk 3]
```diff
@@ -164,3 +174,95 @@ func (n *Node) apiGenToken(c *gin.Context) {
...
+		}
+		return true
+	}
+
+	jwtAttribute := n.nodeAttributes.Filter(predicate)
+	if len(jwtAttribute) != 1 {
+		err := errors.New("Node: apiGenerateGuestToken: got more than 1 jwt_key")
+		api.AbortRequest(c, http.StatusInternalServerError, "multiple_jwt_keys", err, n.log)
+		return
+	}
```
Comment (cnaize) [inline universe/node/api_auth.go:206]: what all this stuff about?

File: universe/node/api_auth.go [hunk 4]
```diff
@@ -164,3 +174,95 @@ func (n *Node) apiGenToken(c *gin.Context) {
...
+
+	var jwtAttributeID entry.AttributeID
+	for k, _ := range jwtAttribute {
+		jwtAttributeID = k
+	}
```
Comment (cnaize) [inline universe/node/api_auth.go:211]: what does it mean?

File: universe/node/api_auth.go [hunk 5]
```diff
@@ -164,3 +174,95 @@ func (n *Node) apiGenToken(c *gin.Context) {
...
+
+	jwtAttribute := n.nodeAttributes.Filter(predicate)
+	if len(jwtAttribute) != 1 {
+		err := errors.New("Node: apiGenerateGuestToken: got more than 1 jwt_key")
+		api.AbortRequest(c, http.StatusInternalServerError, "multiple_jwt_keys", err, n.log)
+		return
+	}
+
+	var jwtAttributeID entry.AttributeID
```
Comment (cnaize) [inline universe/node/api_auth.go:208]: why you use empty attribute id?

File: universe/node/api_auth.go [hunk 6]
```diff
@@ -164,3 +174,95 @@ func (n *Node) apiGenToken(c *gin.Context) {
...
+	if err != nil {
+		err = errors.WithMessage(err, "Node: apiGenerateGuestToken: failed to generate JWT tokens pair")
+		api.AbortRequest(c, http.StatusInternalServerError, "failed_to_generate_jwt", err, n.log)
+		return
+	}
+
+	uid := uuid.Nil
+	if inQuery.UserID != "" {
+		uid = uuid.MustParse(inQuery.UserID)
+	}
```
Comment (cnaize) [inline universe/node/api_auth.go:238]: if you'll create user with nil id, you broke our lock unity logic

File: universe/node/api_users.go [hunk 7]
```diff
@@ -244,82 +235,12 @@ func (n *Node) apiGetOrCreateUserFromTokens(c *gin.Context, accessToken, idToken
 
 	// TODO: check issuer
 
-	nodeSettings, ok := n.GetNodeAttributeValue(
+	_, ok := n.GetNodeAttributeValue(
 		entry.NewAttributeID(universe.GetSystemPluginID(), universe.Attributes.Node.Settings.Name),
 	)
 	if !ok {
 		return nil, http.StatusInternalServerError, errors.Errorf("failed to get node settings")
 	}
```
Comment (cnaize) [inline universe/node/api_users.go:243]: why does you use this check?

File: universe/node/api_users.go [hunk 8]
```diff
@@ -244,82 +235,12 @@ func (n *Node) apiGetOrCreateUserFromTokens(c *gin.Context, accessToken, idToken
...
```
Comment (cnaize) [inline universe/node/api_users.go:244]: why you just dropped all creation user logic?

File: universe/common/auth/jwt.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+		"auth":    JWT,
+		"refresh": rtJWT,
+	}
+
+	return tokensMap, nil
+}
+
+// ValidateToken checks if a given signed token
+// string is a still valid JWT.
+func ValidateToken(jwt *JWTClaim) (string, error) {
```
Comment (cnaize) [inline universe/common/auth/jwt.go:72]: do you actually mark all self signed tokens with valid expire as valid?

**Review by cnaize — CHANGES_REQUESTED** [review]

File: universe/common/auth/jwt.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+)
+
+type JWTClaim struct {
+	UserID string `json:"userID"`
+	jwt.StandardClaims
+	SignedString string `json:"signedString"`
+}
+
+// GenerateJWTPair generates a JWT auth token and a refresh token pair.
+func GenerateJWTPair(userID string, secret []byte) (map[string]*JWTClaim, error) {
```
Comment (cnaize) [inline universe/common/auth/jwt.go:20]: return new jwt token instead of pair

File: universe/common/auth/jwt.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+// ValidateToken checks if a given signed token
+// string is a still valid JWT.
+func ValidateToken(jwt *JWTClaim) (string, error) {
+	if jwt.ExpiresAt < time.Now().Local().Unix() {
+		return "", errors.New("token expired")
+	}
+	return jwt.SignedString, nil
+}
+
+func GetJWTClaimFromContext(c *gin.Context) (*JWTClaim, error) {
```
Comment (cnaize) [inline universe/common/auth/jwt.go:79]: supposed to be token instead of claim

**Review by cnaize — CHANGES_REQUESTED** [review]

File: universe/common/api/middleware/verify_user.go [hunk 12]
```diff
@@ -8,11 +8,18 @@ import (
 	"go.uber.org/zap"
 
 	"github.com/momentum-xyz/ubercontroller/universe/common/api"
+	"github.com/momentum-xyz/ubercontroller/universe/common/auth"
 )
 
 func VerifyUser(log *zap.SugaredLogger) gin.HandlerFunc {
 	return func(c *gin.Context) {
-		token, err := api.VerifyToken(c, api.GetTokenFromRequest(c))
+		claim, err := auth.GetJWTClaimFromContext(c)
```
Comment (cnaize) [inline universe/common/api/middleware/verify_user.go:16]: who supposed to store token where?

**Comment by boyan-d** [comment]
Closing this in favor of

**Pull request closed without merging**